iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp

```diff
@@ -64,11 +64,13 @@ class RelativePointer : public BaseRelativePointer
     template <typename U = T>
     typename std::enable_if<!std::is_void<U>::value, const U&>::type operator*() const noexcept;
 
-    /// @brief read-only access to the underlying object
+    /// @brief read-only access to the underlying object. If the RelativePointer does not point to anything the
+    /// application might terminate.
     /// @return a pointer to the underlying object
     T* operator->() const noexcept;
```

Comment:
This breaks the const correctness and it should be 
`const T* operator->() const noexcept;`
Furthermore, we should introduce a `T* operator->() noexcept` to provide write access.

I know that the STL `unique_ptr` and `shared_ptr` does it differently but here I totally disagree with their behavior. I know that the relative pointer does not own the memory but it provide access to it. When a function would like to have a pointer as argument one can make sure with:
```cxx
void myFunc(const int * ptr);
```
That it is not changed but in here it would mean to generate a completely new type `RelativePointer<const T>` where the memory is always const. But most likely users will misuse it quickly and assume:
```cxx
void myFunc(const RelativePointer<int> ptr);
```
would assume that `myFunc` cannot change the content when it actually can.

See also the `cxx::unique_ptr` where we already ensure the const correctness with the proposal from above.

Reply:
@elfenpiff I know what you mean, but I still disagree here. I believe a lot of effort has gone into designing the STL APIs and it's also quite proven in use. We might not even be aware for certain corner cases.

I would suggest to leave it as it is and also adapt it accordingly in `cxx::unique_ptr` as the user can control the constness like:

```
RelativePointer<T> p             ===> T* p
const RelativePointer<T> p       ===> T* const p
RelativePointer<const T> p       ===> const T* p
const RelativePointer<const T> p ===> const T* const p
```

To prevent the potential misuse, I would suggest to add the above to doxygen. What do you think?

Reply:
@mossmaurice This is the intention and consistent with other fancy pointer APIs to my knowledge, so we should keep it like this. Adding the comment would help.

@elfenpiff The user assumption would be wrong then and she should use 

```
void myFunc(RelativePointer<const int> ptr);
```
The system is consistent and documentation is available.

Another change I would make is revert the name to its original `relative_ptr` for STL style consistency if possible given the current tooling (was like this before some changes before this PR).